core/src/main/java/com/scalar/db/util/groupcommit/GroupManager.java

```diff
@@ -211,4 +216,15 @@ int sizeOfNormalGroupMap() {
   int sizeOfDelayedGroupMap() {
     return delayedGroupMap.size();
   }
+
+  void abortAllGroups() {
```

Comment:
Currently, we don't explicitly abort ongoing transactions when closing transaction manager instances.

> What if the application recreates the transaction manager including the GroupCommitter while there are failed in-flight transactions? The new GroupCommitter doesn't have any information about the in-flight transactions, resulting in that no one aborts them and the clients will wait forever.

Sorry, I don't fully understand why the clients will wait forever. This means that the new GroupCommitter instance interferes the existing GroupCommitter instance? I thought the GroupCommitter instances independently ran.

Reply:
Yeah, this issue is a very uncommon and complicated... Actually, new GroupCommitter doesn't matter in this case. Maybe the following sample code which reproduces the issue is helpful.

```java
 Properties properties = new Properties();
  ...
  properties.put("scalar.db.consensus_commit.coordinator.group_commit.enabled", "true");
  // This is too large, but emulating that a lot of groups are contained in DelayedSlotMoveWorker.
  properties.put("scalar.db.consensus_commit.coordinator.group_commit.delayed_slot_move_timeout_millis", "10000");
  // The default value is 60 seconds. Shoten this to easily to see if the issue happens.
  properties.put("scalar.db.consensus_commit.coordinator.group_commit.old_group_abort_timeout_seconds", "20");
  TransactionFactory factory = TransactionFactory.create(properties);
  DistributedTransactionManager transactionManager = factory.getTransactionManager();
  ExecutorService executorService = Executors.newCachedThreadPool();

  // Transaction#1 should fail in the execution phase.
  Future<?> future1 = executorService.submit(() -> {
    DistributedTransaction txn;
    try {
      txn = transactionManager.begin();
    } catch (TransactionException e) {
      throw new RuntimeException(e);
    }

    try {
      txn.get(
          Get.newBuilder()
              .namespace("ycsb")
              .table("usertable")
             // Wrong key name to fail in the execution phase.
              .partitionKey(Key.ofInt("xxxxxxx_ycsb_key", 42))
              .build());
      txn.commit();
    } catch (TransactionException e) {
      // Don't abort the transaction.
      throw new RuntimeException(e);
    }
  });

  // Transaction#2 should be suspended in the commit phase.
  Future<?> future2 = executorService.submit(() -> {
    DistributedTransaction txn;
    try {
      txn = transactionManager.begin();
    } catch (TransactionException e) {
      throw new RuntimeException(e);
    }

    try {
      Optional<Result> result = txn.get(
          Get.newBuilder()
              .namespace("ycsb")
              .table("usertable")
              .partitionKey(Key.ofInt("ycsb_key", 42))
              .build());
      txn.commit();
      return result;
    } catch (TransactionException e) {
      throw new RuntimeException(e);
    }
  });

  executorService.shutdown();

  TimeUnit.SECONDS.sleep(2);
  // At this moment, the group contains these slots
  // - Slot#1 (transaction#1)
  //   - No value (Snapshot) is set since the transaction failed before committing
  //   - If the client aborts the transaction, this slot is removed and Slot#2 can proceed.
  //   - Also, in normal case, this should be moved to a DelayedGroup (or aborted) by DelayedSlotMoveWorker,
  //     and Slot#2 can move forward. But the worker can be delayed and the user can close the transaction manager before that
  // - Slot#2 (transaction#2)
  //   - A value (Snapshot) is set. And the client thread is waiting until the group is committed

  transactionManager.close();
  // All the background workers are closed. No one is taking care of the information of in-flight groups.

  try {
    future1.get();
  } catch (ExecutionException e) {
    // Expected
    System.out.println("Caught an exception from the first transaction expectedly: " + e.getMessage());
  } catch (InterruptedException e) {
    throw new RuntimeException(e);
  }

  try {
    long start = System.currentTimeMillis();
    // This gets stuck unfortunately forever even waiting more than
    // `old_group_abort_timeout_seconds` since no background worker is running.
    future2.get();
    System.out.println("Got the result from the second transaction. Duration: " + (System.currentTimeMillis() - start));
  } catch (ExecutionException | InterruptedException e) {
    throw new RuntimeException(e);
  }

  // transactionManager.close();
```

Without this PR's change, the above example code lets the client thread for transaction#2 stuck forever.

Reply:
Thank you for the explanation. I may not fully understand the issue yet, but I think we should wait for the queues for the worker threads to be empty when closing GroupCommitter. What do you think?

Reply:
IMO, if users close the transaction manager instance even though there are ongoing transactions, I believe that responsibility lies with the users. Therefore, I think we can simply go with option #2.

Reply:
Yeah, that sounds reasonable. The option 3 I suggested was just based on the option 2 by adding the timeout. We can incrementally add the timeout mechanism to the option 2 later if needed.

But, on second thougtht, I just noticed what if clients keep starting new transactions which I faced in https://github.com/scalar-labs/scalar-jepsen. In this case, probably it won't finish and the test will finally fail. So, let me go with option 3 just in case.

Reply:
Hmm. The both option 2 and 3 should make GroupCommitter reject new requests (i.e., `reserve()`) once it's closed. With this logic, option 2 also can handle the situation I mentioned in the previous comment. I'll take the simpler option 2 😅